[MLIR] [Vector] Added canonicalizer for folding from_elements + transpose

keshavvinayak01 opened the pull request:
## Description
Adds a new canonicalizer that folds `vector.from_elements(vector.transpose))` => `vector.from_elements`. This canonicalization reorders the input elements for `vector.from_elements`, adjusts the output shape to match the effect of the transpose op and eliminating its need.

## Testing
Added a 2D vector lit test that verifies the working of the rewrite.

keshavvinayak01 committed:
Added canonicalization (vector.from_elements + vector.transpose -> vector.transpose)

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Formatted

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

banach-space reviewed (commented):
Thanks!

keshavvinayak01 commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -6723,6 +6723,61 @@ class FoldTransposeShapeCast final : public OpRewritePattern<TransposeOp> {
+
+    // For each element in destination row-major order, pick the corresponding
+    // source element.
+    for (int64_t lin = 0; lin < dstNumElements; ++lin) {
```
So "lin" is short for "linear index" - it's a 1D index that represents the position of an element when the multi-dimensional vector is laid out in row-major order in memory. I felt it was a good iter name.

keshavvinayak01 committed:
Addressed comments:
1. Minor nitpicks in code formatting.
2. More lit tests, convering 1D, 2D, 3D cases.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

banach-space commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -308,6 +308,18 @@ func.func @constant_mask_transpose_to_transposed_constant_mask() -> (vector<2x3x
+  %v = vector.from_elements %a0, %a1, %a2, %a3, %a4, %a5 : vector<2x3xi32>
+  %t = vector.transpose %v, [1, 0] : vector<2x3xi32> to vector<3x2xi32>
+  return %t : vector<3x2xi32>
+  // CHECK: %[[R:.*]] = vector.from_elements %arg0, %arg3, %arg1, %arg4, %arg2, %arg5 : vector<3x2xi32>
```
This comment has not been resolved, so "unresolved".

In case this was not clear, we shouldn't rely on "actual" variable names (e.g. `%arg0`) in FileCheck `CHECK` lines - these can change in the future. Instead, please use FileCheck variables (e.g. `%[[ARG0:.*]]` or similar).

Similar comment for other tests.

banach-space commented on mlir/test/Dialect/Vector/canonicalize.mlir:
```diff
@@ -308,6 +308,41 @@ func.func @constant_mask_transpose_to_transposed_constant_mask() -> (vector<2x3x
+// CHECK-LABEL: transpose_from_elements_3d
+func.func @transpose_from_elements_3d(
+  %arg0: i32, %arg1: i32, %arg2: i32, %arg3: i32, %arg4: i32, %arg5: i32,
+  %arg6: i32, %arg7: i32, %arg8: i32, %arg9: i32, %arg10: i32, %arg11: i32
```
Kind request - use variable names that make it easy to track the actual transposition. For example:
```suggestion
  %el_0_0_0: i32, %el_0_0_1: i32, %el_0_0_2: i32, %el_0_1_0: i32, ...
```

keshavvinayak01 replied:
Good suggestion! will do.

keshavvinayak01 committed:
Explainable arg names in lit test

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Addressed Comments:
1. Changed variable name of linearIdx iterator.
2. Moved canonicalizer lit tests to other vector.from_elements tests.
3. Added blocked comments signaling beginning, end, and name of the pattern.

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

dcaballe commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -6723,6 +6723,61 @@ class FoldTransposeShapeCast final : public OpRewritePattern<TransposeOp> {
 };
 
+/// Folds transpose(from_elements(...)) into a new from_elements with permuted
+/// operands matching the transposed shape.
```
Could you add a before and after IR example? That usually helps a lot with understanding.

keshavvinayak01 replied:
Done, thanks!

dcaballe commented:
LGTM, thanks! (modulo pending comments)

keshavvinayak01 committed:
Added example for folder

Signed-off-by: Keshav Vinayak Jha <keshavvinayakjha@gmail.com>

keshavvinayak01 committed:
Merge branch 'main' into users/keshavvinayak01/from-elements-transpose-fold
